File: decorator-pattern/coffee/coffee.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
// Concrete component
var BasicCoffee = /** @class */ (function () {
    function BasicCoffee() {
    }
    BasicCoffee.prototype.cost = function () {
        return 5;
    };
    BasicCoffee.prototype.description = function () {
        return "Basic coffee";
    };
    return BasicCoffee;
}());
// Decorator (base)
var CoffeeDecorator = /** @class */ (function () {
    function CoffeeDecorator(c) {
        this.coffee = c;
    }
    return CoffeeDecorator;
}());
// --------- Concrete Decorators -------------
var MilkDecorator = /** @class */ (function (_super) {
    __extends(MilkDecorator, _super);
    function MilkDecorator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MilkDecorator.prototype.cost = function () {
        return this.coffee.cost() + 1.5;
    };
    MilkDecorator.prototype.description = function () {
        return this.coffee.description() + ", Milk";
    };
    return MilkDecorator;
}(CoffeeDecorator));
var SugarDecorator = /** @class */ (function (_super) {
    __extends(SugarDecorator, _super);
    function SugarDecorator() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SugarDecorator.prototype.cost = function () {
        return this.coffee.cost() + 0.5;
    };
    SugarDecorator.prototype.description = function () {
        return this.coffee.description() + ", Sugar";
    };
    return SugarDecorator;
}(CoffeeDecorator));
// Usage
var coffee = new BasicCoffee();
console.log("".concat(coffee.description(), " -- ").concat(coffee.cost()));
var milkCoffee = new MilkDecorator(coffee);
console.log("".concat(milkCoffee.description(), " -- ").concat(milkCoffee.cost()));
var milkAndSugarCoffee = new SugarDecorator(milkCoffee);
console.log("".concat(milkAndSugarCoffee.description(), " -- ").concat(milkAndSugarCoffee.cost()));

File: decorator-pattern/coffee/coffee.rs
// Component trait
trait Coffee {
    fn cost(&self) -> f64;
    fn description(&self) -> String;
}

// Concrete component
struct BasicCoffee;

impl Coffee for BasicCoffee {
    fn cost(&self) -> f64 {
        10.0
    }

    fn description(&self) -> String {
        "A basic coffee".to_string()
    }
}

// Decorator trait
// struct CoffeeDecorator<T: Coffee> {
//     coffee: T,
// }

// impl<T: Coffee> CoffeeDecorator<T> {
//     fn new(coffee: T) -> Self {
//         Self { coffee }
//     }
// }

// Concrete decorators
struct MilkDecorator<T: Coffee> {
    coffee: T,
}

impl<T: Coffee> MilkDecorator<T> {
    fn new(coffee: T) -> Self {
        Self { coffee }
    }
}

impl<T: Coffee> Coffee for MilkDecorator<T> {
    fn cost(&self) -> f64 {
        self.coffee.cost() + 5.0
    }

    fn description(&self) -> String {
        format!("{}, with milk", self.coffee.description())
    }
}

struct SugarDecorator<T: Coffee> {
    coffee: T,
}

impl<T: Coffee> SugarDecorator<T> {
    fn new(coffee: T) -> Self {
        Self { coffee }
    }
}

impl<T: Coffee> Coffee for SugarDecorator<T> {
    fn cost(&self) -> f64 {
        self.coffee.cost() + 1.0
    }

    fn description(&self) -> String {
        format!("{}, with sugar", self.coffee.description())
    }
}

fn main() {
    let basic_coffee = BasicCoffee;
    println!("{} -> ${}", basic_coffee.description(), basic_coffee.cost());

    let milk_coffee = MilkDecorator::new(basic_coffee);
    println!("{} -> ${}", milk_coffee.description(), milk_coffee.cost());

    let milk_sugar_coffee = SugarDecorator::new(milk_coffee);
    println!(
        "{} -> ${}",
        milk_sugar_coffee.description(),
        milk_sugar_coffee.cost()
    );
}
